recirq/quantum_chess/quantum_board.py

```diff
@@ -331,16 +334,81 @@ def _generate_accumulations(self, repetitions: int = 1000) -> None:
+            helper_board = CirqBoard(self.init_basis_state, self.sampler,
+                                     self.device, self.error_mitigation,
+                                     self.noise_mitigation,
+                                     self.transformer if self.device else None)
+            sample_jump_move = move.Move('b1',
+                                         'c3',
+                                         target2='a3',
+                                         move_type=enums.MoveType.SPLIT_JUMP,
+                                         move_variant=enums.MoveVariant.BASIC)
+            helper_board.do_move(sample_jump_move)
+            probs = helper_board.get_probability_distribution(
+                cache_key.repetitions, use_cache=False)
+            self.cache[cache_key] = {
+                "source": 0,
+                "target": probs[square_to_bit(sample_jump_move.target)],
+                "target2": probs[square_to_bit(sample_jump_move.target2)]
+            }
+            self.debug_log += helper_board.debug_log
+
+    @staticmethod
+    def _apply_cache(probability, m, cache_value):
+        for k, v in cache_value.items():
+            square = getattr(m, k)
+            probability[square_to_bit(square)] = v
+        return probability
+
     def get_probability_distribution(self,
-                                     repetitions: int = 1000) -> List[float]:
+                                     repetitions: int = 1000,
+                                     use_cache=True) -> List[float]:
         """Returns the probability of a piece being in each square.
 
         The values are returned as a list in the same ordering as a
         bitboard.
         """
+        move_history_cache_key = (tuple(self.move_history), repetitions)
+        move_history_old_cache_key = (tuple(self.move_history[:-1]),
+                                      repetitions)
+        if use_cache and self.move_history:
+            last_move = self.move_history[-1]
+            if move_history_cache_key in self.move_history_probabilities_cache:
+                return self.move_history_probabilities_cache[
+                    move_history_cache_key]
+            cache_key = cache_key_from_move(last_move, repetitions)
+            if move_history_old_cache_key in self.move_history_probabilities_cache and cache_key in self.cache:
+                probs = self._apply_cache(
+                    self.move_history_probabilities_cache[
+                        move_history_old_cache_key], last_move,
+                    self.cache[cache_key])
+                self.probabilities = probs
+                self.move_history_probabilities_cache[
+                    move_history_cache_key] = probs.copy()
+                del self.cache[cache_key]
+                return probs
+
         if self.accumulations_repetitions != repetitions:
             self._generate_accumulations(repetitions)
 
+        self.move_history_probabilities_cache[
```

Comment:
How about making this an array instead of having lists of moves as dictionary keys?

The array could be the same length as `move_history` and the entries a 2-tuple of (repetitions, probabilities) (with -1 for repetitions if the circuit was never run for that move).

Then you could just check if the entry at index -2 has enough repetitions. (IIUC, this PR is only implementing the "Generic Moves" variant as described in the doc so the move lists don't matter except to see whether we know the probabilities after the second-to-last move.)

Reply:
Yeah I think it's rare enough to ignore. In general I don't think the behavior of matching an exact number of repetitions makes sense; it would be better to use a cached value if it has *at least* as many samples as requested. And then we could store just the largest sample... that's out of scope for this PR of course :)

Reply:
Done.